File: scripts/generate_markdown_mirrors.py
#!/usr/bin/env python3
"""
Generate Markdown mirrors (.md) for all HTML pages
Optimized for AI crawler readability
"""

import os
import re
import argparse
from pathlib import Path
from datetime import datetime

try:
    from bs4 import BeautifulSoup
    from markdownify import markdownify as md
except ImportError:
    print("Error: Required packages not installed.")
    print("Please run: pip3 install beautifulsoup4 markdownify")
    exit(1)

# Strip-Regeln: Elemente die entfernt werden
STRIP_TAGS = ['nav', 'footer', 'script', 'style', 'noscript', 'iframe', 'svg', 'form', 'button']

# Class-Muster die entfernt werden (Navigation, Widgets, etc.)
STRIP_CLASS_PATTERNS = [
    r'nav', r'footer', r'header', r'menu',
    r'cta-split', r'cookie', r'consent',
    r'chat', r'widget', r'intercom', r'crisp', r'drift',
    r'tally', r'hubspot', r'ghl',
    r'sticky', r'announcement', r'newsletter',
    r'booking-modal'  # Zahnarztpraxis-spezifisch
]

def should_strip_element(element):
    """Check if element should be removed"""
    # Tag-basiert
    if element.name in STRIP_TAGS:
        return True

    # Class-basiert
    if element.get('class'):
        classes = ' '.join(element.get('class'))
        for pattern in STRIP_CLASS_PATTERNS:
            if re.search(pattern, classes, re.IGNORECASE):
                return True

    # ID-basiert (zusätzlich)
    if element.get('id'):
        element_id = element.get('id')
        for pattern in STRIP_CLASS_PATTERNS:
            if re.search(pattern, element_id, re.IGNORECASE):
                return True

    return False

def clean_html(html_content):
    """Remove navigation, footer, scripts, etc."""
    soup = BeautifulSoup(html_content, 'html.parser')

    # Remove elements
    for element in soup.find_all():
        if should_strip_element(element):
            element.decompose()

    return str(soup)

def html_to_markdown(html_content, page_url, title=""):
    """Convert HTML to clean Markdown"""
    # Clean HTML first
    cleaned = clean_html(html_content)

    # Convert to Markdown
    markdown = md(cleaned, heading_style="ATX", bullets="-")

    # Clean up excessive newlines
    markdown = re.sub(r'\n{3,}', '\n\n', markdown)

    # Add frontmatter
    frontmatter = f"""---
title: {title}
url: {page_url}
last_updated: {datetime.now().strftime('%Y-%m-%d')}
---

"""

    return frontmatter + markdown.strip()

def generate_mirrors(site_root, base_url, skip_paths=None):
    """Generate .md mirrors for all .html files"""
    if skip_paths is None:
        skip_paths = ['/api/', '/app/', '/dashboard/', '/login/']

    site_root = Path(site_root)
    generated = []

    for html_file in site_root.rglob('*.html'):
        # Skip paths
        rel_path = str(html_file.relative_to(site_root))
        if any(skip in rel_path for skip in skip_paths):
            print(f"⊘ Skipping {rel_path} (in skip_paths)")
            continue

        # Read HTML
        with open(html_file, 'r', encoding='utf-8') as f:
            html_content = f.read()

        # Extract title
        soup = BeautifulSoup(html_content, 'html.parser')
        title_tag = soup.find('title')
        title = title_tag.get_text() if title_tag else html_file.stem

        # Generate URL
        page_url = base_url + '/' + rel_path

        # Convert to Markdown
        markdown_content = html_to_markdown(html_content, page_url, title)

        # Write .md file (same directory as .html)
        md_file = html_file.with_suffix('.md')
        with open(md_file, 'w', encoding='utf-8') as f:
            f.write(markdown_content)

        generated.append(str(md_file.relative_to(site_root)))
        print(f"✓ {rel_path} → {md_file.name}")

    return generated

def main():
    parser = argparse.ArgumentParser(description='Generate Markdown mirrors for HTML pages')
    parser.add_argument('--site-root', required=True, help='Site root directory (e.g., .output)')
    parser.add_argument('--base-url', required=True, help='Base URL (e.g., https://zahnarztpraxis-schwabing.de)')

    args = parser.parse_args()

    print(f"Generating Markdown mirrors...")
    print(f"Site root: {args.site_root}")
    print(f"Base URL: {args.base_url}")
    print()

    generated = generate_mirrors(args.site_root, args.base_url)

    print()
    print(f"✅ Generated {len(generated)} Markdown mirrors:")
    for md_file in generated:
        print(f"   - {md_file}")

if __name__ == '__main__':
    main()

File: scripts/generate_markdown_mirrors.mjs
#!/usr/bin/env node
/**
 * Generate Markdown mirrors (.md) for all HTML pages
 * Node.js version - works on Vercel without Python
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { parse } from 'node-html-parser';
import TurndownService from 'turndown';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Elements to strip
const STRIP_TAGS = ['nav', 'footer', 'script', 'style', 'noscript', 'iframe', 'svg', 'form', 'button'];

// Class patterns to strip
const STRIP_CLASS_PATTERNS = [
  /nav/i, /footer/i, /header/i, /menu/i,
  /cta-split/i, /cookie/i, /consent/i,
  /chat/i, /widget/i, /intercom/i, /crisp/i, /drift/i,
  /tally/i, /hubspot/i, /ghl/i,
  /sticky/i, /announcement/i, /newsletter/i,
  /booking-modal/i
];

function shouldStripElement(element) {
  // Check tag name
  if (STRIP_TAGS.includes(element.tagName?.toLowerCase())) {
    return true;
  }

  // Check classes
  const classList = element.getAttribute('class');
  if (classList) {
    for (const pattern of STRIP_CLASS_PATTERNS) {
      if (pattern.test(classList)) {
        return true;
      }
    }
  }

  // Check ID
  const id = element.getAttribute('id');
  if (id) {
    for (const pattern of STRIP_CLASS_PATTERNS) {
      if (pattern.test(id)) {
        return true;
      }
    }
  }

  return false;
}

function cleanHTML(html) {
  const root = parse(html);

  // Remove unwanted elements
  const allElements = root.querySelectorAll('*');
  for (const element of allElements) {
    if (shouldStripElement(element)) {
      element.remove();
    }
  }

  return root.toString();
}

function htmlToMarkdown(htmlContent, pageUrl, title) {
  // Clean HTML first
  const cleaned = cleanHTML(htmlContent);

  // Convert to Markdown
  const turndownService = new TurndownService({
    headingStyle: 'atx',
    bulletListMarker: '-'
  });

  const markdown = turndownService.turndown(cleaned);

  // Clean up excessive newlines
  const cleanedMd = markdown.replace(/\n{3,}/g, '\n\n');

  // Add frontmatter
  const now = new Date();
  const dateStr = now.toISOString().split('T')[0];

  const frontmatter = `---
title: ${title}
url: ${pageUrl}
last_updated: ${dateStr}
---

`;

  return frontmatter + cleanedMd.trim();
}

function generateMirrors(siteRoot, baseUrl) {
  const generated = [];
  const skipPaths = ['/api/', '/app/', '/dashboard/', '/login/'];

  function processDirectory(dir) {
    const files = fs.readdirSync(dir);

    for (const file of files) {
      const filePath = path.join(dir, file);
      const stat = fs.statSync(filePath);

      if (stat.isDirectory()) {
        processDirectory(filePath);
      } else if (file.endsWith('.html')) {
        const relPath = path.relative(siteRoot, filePath);

        // Skip paths
        if (skipPaths.some(skip => relPath.includes(skip))) {
          console.log(`⊘ Skipping ${relPath}`);
          continue;
        }

        // Read HTML
        const htmlContent = fs.readFileSync(filePath, 'utf-8');

        // Extract title
        const root = parse(htmlContent);
        const titleTag = root.querySelector('title');
        const title = titleTag?.text || path.basename(file, '.html');

        // Generate URL
        const pageUrl = baseUrl + '/' + relPath;

        // Convert to Markdown
        const markdown = htmlToMarkdown(htmlContent, pageUrl, title);

        // Write .md file
        const mdPath = filePath.replace('.html', '.md');
        fs.writeFileSync(mdPath, markdown, 'utf-8');

        generated.push(path.relative(siteRoot, mdPath));
        console.log(`✓ ${relPath} → ${path.basename(mdPath)}`);
      }
    }
  }

  processDirectory(siteRoot);
  return generated;
}

function main() {
  const args = process.argv.slice(2);
  const siteRoot = args.find(a => a.startsWith('--site-root='))?.split('=')[1];
  const baseUrl = args.find(a => a.startsWith('--base-url='))?.split('=')[1];

  if (!siteRoot || !baseUrl) {
    console.error('Usage: node generate_markdown_mirrors.mjs --site-root=<path> --base-url=<url>');
    process.exit(1);
  }

  console.log('Generating Markdown mirrors...');
  console.log(`Site root: ${siteRoot}`);
  console.log(`Base URL: ${baseUrl}`);
  console.log();

  const generated = generateMirrors(siteRoot, baseUrl);

  console.log();
  console.log(`✅ Generated ${generated.length} Markdown mirrors:`);
  for (const file of generated) {
    console.log(`   - ${file}`);
  }
}

main();
